File: packages/sdk-php/src/TeamAPI.php
<?php

declare(strict_types=1);

namespace Melaya;

/**
 * Team API — manage project team membership, roles, and invitations.
 *
 * Maps to /api/v1/private/projects/:project/members/* and /api/v1/private/team/*.
 *
 * @example
 * ```php
 * $sdk = new Melaya\Melaya(apiKey: getenv('MK'));
 *
 * $members = $sdk->team->listMembers('my-project');
 * $sdk->team->invite('my-project', 'alice');
 *
 * $link = $sdk->team->createInviteLink('my-project');
 * // Share $link['inviteLink'] with new team members
 * ```
 */
class TeamAPI
{
    public function __construct(private readonly HttpClient $http) {}

    /** List members of a project team. */
    public function listMembers(string $project): array
    {
        return $this->http->get('/api/v1/private/projects/' . rawurlencode($project) . '/members');
    }

    /**
     * Invite a user to a project team by username.
     *
     * @param string $project  Project name
     * @param string $username The username to invite
     */
    public function invite(string $project, string $username): array
    {
        return $this->http->post(
            '/api/v1/private/projects/' . rawurlencode($project) . '/members/invite',
            ['username' => $username]
        );
    }

    /**
     * Create a shareable invite link for a project.
     * Returns ['ok' => true, 'inviteLink' => '...'].
     */
    public function createInviteLink(string $project): array
    {
        return $this->http->post('/api/v1/private/projects/' . rawurlencode($project) . '/invite-link');
    }

    /**
     * Accept a project invite using the token from an invite link.
     *
     * @param string $token The invite token from the invite URL
     */
    public function acceptInvite(string $token): array
    {
        return $this->http->post('/api/v1/private/team/invite/accept', ['token' => $token]);
    }

    /**
     * Update a team member's role in a project.
     *
     * @param string $project Project name
     * @param string $userId  The user's ID
     * @param string $role    One of 'owner' | 'editor' | 'viewer'
     */
    public function updateMemberRole(string $project, string $userId, string $role): array
    {
        return $this->http->patch(
            '/api/v1/private/projects/' . rawurlencode($project) . '/members/' . rawurlencode($userId),
            ['role' => $role]
        );
    }

    /** Remove a member from a project team. */
    public function removeMember(string $project, string $userId): array
    {
        return $this->http->delete(
            '/api/v1/private/projects/' . rawurlencode($project) . '/members/' . rawurlencode($userId)
        );
    }

    // ── Pipeline visibility ──────────────────────────────────────────────────

    /** Get visibility settings for a pipeline within a project. */
    public function getPipelineVisibility(string $project, string $pipeline): array
    {
        return $this->http->get(
            '/api/v1/private/projects/' . rawurlencode($project) . '/pipelines/' . rawurlencode($pipeline) . '/visibility'
        );
    }

    /**
     * Set pipeline visibility within a project.
     *
     * @param array $body Visibility settings
     */
    public function setPipelineVisibility(string $project, string $pipeline, array $body): array
    {
        return $this->http->put(
            '/api/v1/private/projects/' . rawurlencode($project) . '/pipelines/' . rawurlencode($pipeline) . '/visibility',
            $body
        );
    }
}

File: packages/sdk-java/src/main/java/org/melaya/TeamAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Map;

import static org.melaya.MarketAPI.params;

/**
 * Team API — manage project team membership, roles, and invitations.
 *
 * <p>Maps to {@code /api/v1/private/projects/:project/members/*} and
 * {@code /api/v1/private/team/*}.
 *
 * @example
 * <pre>{@code
 * JsonNode members = melaya.team().listMembers("my-project");
 * melaya.team().invite("my-project", "alice");
 * JsonNode link = melaya.team().createInviteLink("my-project");
 * }</pre>
 */
public class TeamAPI {

    private final HttpClient http;

    TeamAPI(HttpClient http) {
        this.http = http;
    }

    // ── Members ───────────────────────────────────────────────────────────────

    /**
     * List members of a project team.
     *
     * @param project the project name
     */
    public JsonNode listMembers(String project) {
        return http.get(
                "/api/v1/private/projects/" + encode(project) + "/members",
                null);
    }

    /**
     * Invite a user to a project team by username.
     *
     * @param project  the project name
     * @param username the username to invite
     */
    public JsonNode invite(String project, String username) {
        return http.post(
                "/api/v1/private/projects/" + encode(project) + "/members/invite",
                params("username", username));
    }

    /**
     * Create a shareable invite link for a project.
     * Returns the URL to share with new team members.
     *
     * @param project the project name
     */
    public JsonNode createInviteLink(String project) {
        return http.post(
                "/api/v1/private/projects/" + encode(project) + "/invite-link",
                null);
    }

    /**
     * Accept a project invite using the token from an invite link.
     *
     * @param token the invite token from the link
     */
    public JsonNode acceptInvite(String token) {
        return http.post("/api/v1/private/team/invite/accept", params("token", token));
    }

    /**
     * Update a team member's role in a project.
     *
     * @param project the project name
     * @param userId  the member's user ID
     * @param role    the new role ({@code "owner"}, {@code "editor"}, or {@code "viewer"})
     */
    public JsonNode updateMemberRole(String project, String userId, String role) {
        return http.patch(
                "/api/v1/private/projects/" + encode(project) + "/members/" + encode(userId),
                params("role", role));
    }

    /**
     * Remove a member from a project team.
     *
     * @param project the project name
     * @param userId  the member's user ID
     */
    public JsonNode removeMember(String project, String userId) {
        return http.delete(
                "/api/v1/private/projects/" + encode(project) + "/members/" + encode(userId),
                null);
    }

    // ── Pipeline visibility ───────────────────────────────────────────────────

    /**
     * Get visibility settings for a pipeline within a project.
     *
     * @param project  the project name
     * @param pipeline the pipeline name
     */
    public JsonNode getPipelineVisibility(String project, String pipeline) {
        return http.get(
                "/api/v1/private/projects/" + encode(project)
                        + "/pipelines/" + encode(pipeline) + "/visibility",
                null);
    }

    /**
     * Set pipeline visibility within a project.
     *
     * @param project  the project name
     * @param pipeline the pipeline name
     * @param body     visibility config map
     */
    public JsonNode setPipelineVisibility(String project, String pipeline, Map<String, Object> body) {
        return http.put(
                "/api/v1/private/projects/" + encode(project)
                        + "/pipelines/" + encode(pipeline) + "/visibility",
                body);
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static String encode(String s) {
        try {
            return java.net.URLEncoder.encode(s, java.nio.charset.StandardCharsets.UTF_8);
        } catch (Exception e) {
            return s;
        }
    }
}

